src/util/vec2d.rs

```diff
@@ -0,0 +1,137 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous `Vec`
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Box<[T]>,
+    cols: usize,
+}
+
+impl<T> Vec2D<T> {
+    /// Initialize a grid of size (`rows`, `cols`) with the given data element.
+    pub fn init(data: T, size: (usize, usize)) -> Vec2D<T>
+    where
+        T: Clone,
+    {
+        let (rows, cols) = size;
+        let len = rows
+            .checked_mul(cols)
+            .unwrap_or_else(|| panic!("{} rows by {} cols exceeds usize::MAX", rows, cols));
```

Comment:
`unwrap_or` + `panic` can be replaced by `expect` instead (with `format!` here as the message isn't constant).

Reply:
`unwrap_or_else` + `panic` is what clippy suggests to do to avoid `format!` being eagerly evaluated. There is unfortunatly not an `expect_or_else` function available.

```
error: use of `expect` followed by a function call
  --> src/util/vec2d.rs:[17](https://github.com/gendx/lzma-rs/runs/7986381449?check_suite_focus=true#step:4:18):42
   |
17 |           let len = rows.checked_mul(cols).expect(&format!(
   |  __________________________________________^
[18](https://github.com/gendx/lzma-rs/runs/7986381449?check_suite_focus=true#step:4:19) | |             "{} rows by {} cols exceeds usize::MAX",
[19](https://github.com/gendx/lzma-rs/runs/7986381449?check_suite_focus=true#step:4:20) | |             rows, cols
[20](https://github.com/gendx/lzma-rs/runs/7986381449?check_suite_focus=true#step:4:21) | |         ));
   | |__________^
   |
   = note: `-D clippy::expect-fun-call` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expect_fun_call
help: try this
   |
17 ~         let len = rows.checked_mul(cols).unwrap_or_else(|| panic!("{} rows by {} cols exceeds usize::MAX",
18 ~             rows, cols));
   |
   ```